### PR TITLE
Sofdmic4ch

### PR DESCRIPTION
**Pull request by maruthi-machani**
This pull request will enable dmic with four channel.

### CODE REVIEW

File: tools/topology/topology1/CMakeLists.txt [hunk 1]
```diff
@@ -292,6 +292,7 @@ set(TPLGS
 	"sof-rn-rt5682-rt1019\;sof-rn-rt5682-rt1019-dts\;-DDTS=`DTS'"
 	"sof-rn-rt5682-max98360\;sof-rn-rt5682-max98360"
 	"sof-acp-rmb\;sof-acp-rmb"
+	 "sof-acp-rmb-dmic4ch\;sof-acp-rmb-dmic4ch"
```
Comment (lyakh) [inline tools/topology/topology1/CMakeLists.txt:295]: the space after the TAB is unneeded

File: src/schedule/dma_multi_chan_domain.c [hunk 2]
```diff
@@ -280,7 +280,7 @@ static bool dma_multi_chan_domain_is_pending(struct ll_schedule_domain *domain,
 	int j;
 
 	for (i = 0; i < dma_domain->num_dma; ++i) {
-		for (j = 0; j < dmas[i].plat_data.channels; ++j) {
+		for (j = 0; (j < dmas[i].plat_data.channels) && ((&dmas[i].chan[j]) != NULL); ++j) {
```
Comment (lyakh) [inline src/schedule/dma_multi_chan_domain.c:283]: this doesn't make sense - `dmas` is a local variable, as long as your stack is fine, address of that expression above will never be `NULL`

File: src/drivers/amd/rembrandt/acp_dmic_dma.c [hunk 3]
```diff
@@ -78,30 +78,30 @@ static void acp_dmic_dma_channel_put(struct dma_chan_data *channel)
...
 	hs_irer = (acp_hstdm_irer_t)io_reg_read((PU_REGISTER_BASE + ACP_HSTDM_IRER));
 	acp_pdm_en = (uint32_t)io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE);
 
-	if (!hs_iter.bits.hstdm_txen && !hs_irer.bits.hstdm_rx_en && !acp_pdm_en)
+	if (!hs_iter.bits.hstdm_txen && !hs_irer.bits.hstdm_rx_en && !acp_pdm_en) {
 		/* Request SMU to set aclk to 600 Mhz */
 		acp_change_clock_notify(600000000);
+		clkmux_sel.u32all = io_reg_read(PU_REGISTER_BASE + ACP_CLKMUX_SEL);
+		clkmux_sel.bits.acp_clkmux_sel = 0x3;
+		io_reg_write((PU_REGISTER_BASE + ACP_CLKMUX_SEL), clkmux_sel.u32all);
```
Comment (lyakh) [inline src/drivers/amd/rembrandt/acp_dmic_dma.c:101]: superfluous internal parentheses

File: src/drivers/amd/rembrandt/acp_dmic_dma.c [hunk 4]
```diff
@@ -193,9 +194,13 @@ static int acp_dmic_dma_stop(struct dma_chan_data *channel)
 	hs_irer = (acp_hstdm_irer_t)io_reg_read((PU_REGISTER_BASE + ACP_HSTDM_IRER));
 	acp_pdm_en = (uint32_t)io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE);
 
-	if (!hs_iter.bits.hstdm_txen && !hs_irer.bits.hstdm_rx_en && !acp_pdm_en)
+	if (!hs_iter.bits.hstdm_txen && !hs_irer.bits.hstdm_rx_en && !acp_pdm_en) {
 		/* Request SMU to set aclk to minimum aclk */
 		acp_change_clock_notify(0);
+		clkmux_sel.u32all = io_reg_read(PU_REGISTER_BASE + ACP_CLKMUX_SEL);
+		clkmux_sel.bits.acp_clkmux_sel = 0x0;
+		io_reg_write((PU_REGISTER_BASE + ACP_CLKMUX_SEL), clkmux_sel.u32all);
```
Comment (lyakh) [inline src/drivers/amd/rembrandt/acp_dmic_dma.c:202]: ditto

File: src/drivers/amd/renoir/acp_dmic_dma.c [hunk 5]
```diff
@@ -194,9 +196,13 @@ static int acp_dmic_dma_stop(struct dma_chan_data *channel)
 	sp_irer = (acp_i2stdm_irer_t)io_reg_read((PU_REGISTER_BASE + ACP_I2STDM_IRER));
 	acp_pdm_en = (uint32_t)io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE);
 
-	if (!sp_iter.bits.i2stdm_txen && !sp_irer.bits.i2stdm_rx_en && !acp_pdm_en)
+	if (!sp_iter.bits.i2stdm_txen && !sp_irer.bits.i2stdm_rx_en && !acp_pdm_en) {
 		/* Request SMU to set aclk to minimum aclk */
 		acp_change_clock_notify(0);
+		clkmux_sel.u32all = io_reg_read(PU_REGISTER_BASE + ACP_CLKMUX_SEL);
+		clkmux_sel.bits.acp_clkmux_sel = 0x0;
+		io_reg_write((PU_REGISTER_BASE + ACP_CLKMUX_SEL), clkmux_sel.u32all);
```
Comment (lyakh) [inline src/drivers/amd/renoir/acp_dmic_dma.c:204]: ditto

File: src/drivers/amd/renoir/acp_dmic_dma.c [hunk 6]
```diff
@@ -93,16 +93,17 @@ static int acp_dmic_dma_start(struct dma_chan_data *channel)
 	sp_irer = (acp_i2stdm_irer_t)io_reg_read((PU_REGISTER_BASE + ACP_I2STDM_IRER));
 	acp_pdm_en = (uint32_t)io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE);
 
-	if (!sp_iter.bits.i2stdm_txen && !sp_irer.bits.i2stdm_rx_en && !acp_pdm_en)
+	if (!sp_iter.bits.i2stdm_txen && !sp_irer.bits.i2stdm_rx_en && !acp_pdm_en) {
 		/* Request SMU to set aclk to 600 Mhz */
 		acp_change_clock_notify(600000000);
+		clkmux_sel.u32all = io_reg_read(PU_REGISTER_BASE + ACP_CLKMUX_SEL);
+		clkmux_sel.bits.acp_clkmux_sel = 0x3;
+		io_reg_write((PU_REGISTER_BASE + ACP_CLKMUX_SEL), clkmux_sel.u32all);
```
Comment (lyakh) [inline src/drivers/amd/renoir/acp_dmic_dma.c:101]: here as well

**Review by lyakh — CHANGES_REQUESTED** [review]

File: src/drivers/amd/renoir/acp_dmic_dma.c (hunk 5 above)
Comment (maruthi-machani) [reply src/drivers/amd/renoir/acp_dmic_dma.c]: done

File: src/drivers/amd/rembrandt/acp_dmic_dma.c (hunk 4 above)
Comment (maruthi-machani) [reply src/drivers/amd/rembrandt/acp_dmic_dma.c]: done

File: src/drivers/amd/rembrandt/acp_dmic_dma.c (hunk 3 above)
Comment (maruthi-machani) [reply src/drivers/amd/rembrandt/acp_dmic_dma.c]: done

File: src/schedule/dma_multi_chan_domain.c (hunk 2 above)
Comment (maruthi-machani) [reply src/schedule/dma_multi_chan_domain.c]: done

File: tools/topology/topology1/CMakeLists.txt (hunk 1 above)
Comment (maruthi-machani) [reply tools/topology/topology1/CMakeLists.txt]: done

File: src/drivers/amd/rembrandt/acp_dmic_dai.c [hunk 7]
```diff
@@ -120,7 +135,15 @@ static int acp_dmic_dai_get_hw_params(struct dai *dai,
...
 		params->rate = ACP_DEFAULT_SAMPLE_RATE;
 	}
-	params->channels = ACP_DEFAULT_NUM_CHANNELS;
+	switch (acpdata->dmic_params.pdm_ch) {
+	case 2:
+	case 4:
+		params->channels = acpdata->dmic_params.pdm_ch;
+		break;
+	default:
+		dai_info(dai, "unsupported channels %d", acpdata->dmic_params.pdm_ch);
```
Comment (lyakh) [inline src/drivers/amd/rembrandt/acp_dmic_dai.c:144]: should this be a warning or an error? Similar in other such "unsupported" locations

File: src/drivers/amd/rembrandt/acp_dmic_dai.c [hunk 8]
```diff
@@ -120,7 +135,15 @@ static int acp_dmic_dai_get_hw_params(struct dai *dai,
...
+				 params->channels = ACP_DEFAULT_NUM_CHANNELS;
```
Comment (lyakh) [inline src/drivers/amd/rembrandt/acp_dmic_dai.c:145]: line misaligned

File: src/drivers/amd/renoir/acp_dmic_dai.c [hunk 9]
```diff
@@ -124,7 +139,15 @@ static int acp_dmic_dai_get_hw_params(struct dai *dai,
...
 	}
-	params->channels = ACP_DEFAULT_NUM_CHANNELS;
+	switch (acpdata->dmic_params.pdm_ch) {
+	case 2:
+	case 4:
+		params->channels = acpdata->dmic_params.pdm_ch;
+		break;
+	default:
+		dai_info(dai, "ACP:unsupported channels %d", acpdata->dmic_params.pdm_ch);
+				 params->channels = ACP_DEFAULT_NUM_CHANNELS;
```
Comment (lyakh) [inline src/drivers/amd/renoir/acp_dmic_dai.c:149]: misaligned, makes it difficult to notice, it looks like a `dai_info()` continuation

File: src/drivers/amd/renoir/acp_dmic_dai.c [hunk 10]
```diff
@@ -51,6 +53,19 @@ static inline int acp_dmic_dai_set_config(struct dai *dai, struct ipc_config_dai
...
+	switch (acpdata->dmic_params.pdm_ch) {
+	case 2:
+		pdm_channels.bits.pdm_no_of_channels = 0;
+		break;
+	case 4:
+		pdm_channels.bits.pdm_no_of_channels = 1;
+		break;
+	default:
+		dai_info(dai, "ACP:acp_dmic_set_config unsupported channels");
+		return -EINVAL;
```
Comment (lyakh) [inline src/drivers/amd/renoir/acp_dmic_dai.c:65]: just to confirm - this is intentional, that in some cases invalid channel number causes an error and in others a default value is just used?

File: src/drivers/amd/rembrandt/acp_dmic_dai.c (hunk 7 above)
Comment (maruthi-machani) [reply src/drivers/amd/rembrandt/acp_dmic_dai.c]: changes done

File: src/drivers/amd/renoir/acp_dmic_dai.c (hunk 9 above)
Comment (maruthi-machani) [reply src/drivers/amd/renoir/acp_dmic_dai.c:149]: changes done

File: src/drivers/amd/renoir/acp_dmic_dai.c (hunk 10 above)
Comment (maruthi-machani) [reply src/drivers/amd/renoir/acp_dmic_dai.c:65]: changes done

**Review by lyakh — APPROVED** [review]

**Review by bhiregoudar — APPROVED** [review]

**Review by sunilkumardommati — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**